In [85]:
import pandas as pd
import numpy as np
import datetime

- Question 1

In [86]:
file_of_path = "../Input/Data.xlsx"
df=pd.read_excel(file_of_path,'月度数据',skipfooter=1,index_col=[0],parse_dates=[0],
date_parser=lambda x:pd.to_datetime(x))
df.tail(n = 3)


指标A_月度_平均  指标B_月度_平均  指标C_月度_平均  指标D_月度_平均  指标E_累计_月度_平均  \
YYYYMM                                                                 
2022-07-01  17.405045  27.664043        NaN  44.938516      1.686957   
2022-08-01  -4.642539  62.344467        NaN  59.259213     19.214227   
2022-09-01  53.881234 -16.021362        NaN  60.479478     32.963925   

            指标F_月度_平均  指标G_月度_平均  指标H_月度_平均  指标I_月度_平均  指标J_累计_月度_平均  
YYYYMM                                                                
2022-07-01  27.448304  34.666621  -0.639802        NaN     50.067729  
2022-08-01  19.493692   8.372807  15.332804        NaN     53.055872  
2022-09-01   2.422561  -5.628644  17.709667        NaN    -20.790993

In [87]:
df.index

DatetimeIndex(['2013-11-01', '2013-12-01', '2014-01-01', '2014-02-01',
               '2014-03-01', '2014-04-01', '2014-05-01', '2014-06-01',
               '2014-07-01', '2014-08-01',
               ...
               '2021-12-01', '2022-01-01', '2022-02-01', '2022-03-01',
               '2022-04-01', '2022-05-01', '2022-06-01', '2022-07-01',
               '2022-08-01', '2022-09-01'],
              dtype='datetime64[ns]', name='YYYYMM', length=107, freq=None)

- Question 2


<strong>1.  </strong>请使用你认为的最简洁的方式：重命名数据框df中的各列列名为如下格式["指标A", "指标B", ..., "指标G"]

In [88]:
df.columns=df.columns.str.strip("_月度_平均")
df.tail(n = 5)


指标A        指标B  指标C        指标D     指标E_累计        指标F  \
YYYYMM                                                                   
2022-05-01 -14.515866  50.271447  NaN  49.223324 -13.012325  -4.745951   
2022-06-01 -14.408964  77.351809  NaN  46.006514  77.763122 -18.050054   
2022-07-01  17.405045  27.664043  NaN  44.938516   1.686957  27.448304   
2022-08-01  -4.642539  62.344467  NaN  59.259213  19.214227  19.493692   
2022-09-01  53.881234 -16.021362  NaN  60.479478  32.963925   2.422561   

                  指标G        指标H  指标I     指标J_累计  
YYYYMM                                            
2022-05-01  50.040173 -20.357725  NaN  12.128351  
2022-06-01  65.412750 -13.425748  NaN   9.032872  
2022-07-01  34.666621  -0.639802  NaN  50.067729  
2022-08-01   8.372807  15.332804  NaN  53.055872  
2022-09-01  -5.628644  17.709667  NaN -20.790993

<strong>2.  </strong>请使用你认为的最简洁的方式：统计数据框df 时间索引列YYYYMM中历年的观测月数量（列名依次为年份Year，观测月数量MonthNum）并将统计好的结果赋值给变量temp_1

In [89]:
temp=pd.Series(df.index).groupby(df.index.year).count()
temp_1=pd.DataFrame({"Year":temp.index,"MonthNum":temp.values})
temp_1

Year  MonthNum
0  2013         2
1  2014        12
2  2015        12
3  2016        12
4  2017        12
5  2018        12
6  2019        12
7  2020        12
8  2021        12
9  2022         9

<strong>3.  </strong>
- 统计**各指标**的以下信息
    - 起始时间（第一个非空时间点，列名命名为`BegDt`）
    - 终止时间（最后一个非空时间点，列名命名为`EndDt`）
    - 观测总数（第一个非空时间点 至 最后一个非空时间点中 的月份个数，列名命名为`Obs`）
    - 缺失值个数和相应的个数占比（第一个非空时间点至最后一个非空时间点中缺失值的个数，列名分别命名为`NaNNum`、`NaNPct`）
    - 负值个数、相应的个数占比、和负值对应的时间点（第一个非空时间点至最后一个非空时间点中负数的个数，列名分别命名为`NegNum`、`NegPct`、`NegLst`）
- 并将统计好的结果赋值给变量`df_summary_1`。可参考变量`df_summary_1` 的内容为如下形式（前文所引括号为各列的列名）

In [90]:
BegDt = df.apply(lambda series: series.first_valid_index())
EndDt=df.apply(lambda series: series.last_valid_index())
Obs=(EndDt).dt.to_period('M').astype(int)-BegDt.dt.to_period('M').astype(int)+1
NaNNum=df.apply(lambda series:series[series.first_valid_index():series.last_valid_index()].isna().sum() )
NaNPct=round(NaNNum/Obs*100,2)
NegNum=df.apply(lambda series:(series[series.first_valid_index():series.last_valid_index()]<0).sum() )
NegPct=round(NegNum/Obs*100,2)
index=df.index
df.index=pd.Series(df.index.values).dt.to_period('M')
NegLst=df.apply(lambda series:series[series.first_valid_index():series.last_valid_index()].index
[series[series.first_valid_index():series.last_valid_index()]<0 ].values )
df_summary_1=pd.DataFrame({"BegDt":BegDt,"EndDt":EndDt,"Obs":Obs,"NaNNum":NaNNum,"NaNPct":NaNPct,
"NegNum":NegNum,"NegPct":NegPct,"NegLst":NegLst})
df.index=index
#df_summary_1["NegPct"]=df_summary_1["NegPct"].apply(lambda list: )
df_summary_1

BegDt      EndDt  Obs  NaNNum  NaNPct  NegNum  NegPct  \
指标A    2015-01-01 2022-09-01   93       2    2.15      13   13.98   
指标B    2013-11-01 2022-09-01  107       1    0.93      28   26.17   
指标C    2013-11-01 2019-12-01   74       0    0.00      20   27.03   
指标D    2013-11-01 2022-09-01  107       1    0.93      18   16.82   
指标E_累计 2013-11-01 2022-09-01  107       2    1.87      20   18.69   
指标F    2013-11-01 2022-09-01  107       1    0.93      19   17.76   
指标G    2013-11-01 2022-09-01  107       0    0.00      27   25.23   
指标H    2019-01-01 2022-09-01   45       0    0.00      10   22.22   
指标I    2013-11-01 2020-06-01   80       0    0.00      27   33.75   
指标J_累计 2015-01-01 2022-09-01   93      38   40.86      16   17.20   

                                                   NegLst  
指标A     [2016-08, 2016-09, 2017-02, 2017-03, 2018-08, ...  
指标B     [2013-11, 2013-12, 2014-04, 2014-06, 2014-08, ...  
指标C     [2013-12, 2014-06, 2014-07, 2014-12, 2015-05, ...  
指标D     [2014-01, 2014-02, 2014-07, 2014-09, 2014-12, ...  
指标E_累计  [2013-11, 2014-03, 2014-06, 2015-01, 2015-06, ...  
指标F     [2014-07, 2015-01, 2015-11, 2016-07, 2016-11, ...  
指标G     [2014-01, 2014-03, 2014-07, 2014-08, 2014-12, ...  
指标H     [2019-01, 2019-02, 2019-09, 2019-11, 2021-08, ...  
指标I     [2014-09, 2014-10, 2014-11, 2015-01, 2015-06, ...  
指标J_累计  [2015-04, 2015-06, 2015-10, 2015-11, 2019-07, ...

- Question 3

<strong>1.  </strong>请使用你认为的最简洁的方式：以2015/01/01为起始时间，2021/12/31为终止时间作为时间筛选条件，保留此时间段内的["指标A", "指标B", "指标D", "指标E_累计", "指标F", "指标G", "指标J_累计"]；
<br>
同时，由于"指标J_累计" 的缺失率NaNPct 的值为40.86\% 超过了20\%，故需要通过自动化筛选的规则予以剔除（Tips：通过Question2的结果变量df_summary_1查询得到需要剔除的指标名称为"指标J_累计"），
<br>即最后实际入选的指标为`["指标A", "指标B", "指标D", "指标E_累计", "指标F", "指标G"]`**（该列表不能直接引用，必须通过列表****`["指标A", "指标B", "指标D", "指标E_累计", "指标F", "指标G", "指标J_累计"]`****结合上述的缺失值筛选规则加工得到）**并将筛选后到的数据框赋值给变量`df1`


In [91]:
df1=df.loc[datetime.datetime(2015,1,1):datetime.datetime(2021,12,31):,
["指标A", "指标B", "指标D", "指标E_累计", "指标F", "指标G"]]
df1

指标A        指标B        指标D     指标E_累计        指标F        指标G
YYYYMM                                                                      
2015-01-01  45.434917 -12.905679  24.022993  -6.834541  -5.296895  52.968511
2015-02-01  31.919156  60.259495  14.858841  40.628302  45.376438  37.912412
2015-03-01  78.077235 -15.167676   7.043763  39.775122  53.367137  52.085428
2015-04-01  49.384843  28.609080  34.357724   8.668052  50.794603  39.574971
2015-05-01   7.336568  56.382674  34.768454  34.298297  15.006340  26.645740
...               ...        ...        ...        ...        ...        ...
2021-08-01   0.147898  44.543486  10.557716  29.407784 -16.316181  38.094446
2021-09-01  43.046844  30.111458  49.393706  54.065227  68.291341  41.396381
2021-10-01 -19.659707  29.809855  79.009139  -6.919456   3.515089  54.807505
2021-11-01  46.220272  39.486479   1.283920  74.777369   1.204036  -0.061766
2021-12-01  33.491257 -11.082423  41.962053  33.863632  33.335368  31.786463

[84 rows x 6 columns]

<strong>2.  </strong>请使用你认为的最简洁的方式：统计变量df1各指标的观测数（count）、均值（mean）、标准差（std）、最小值（min）、25\%分位数（25%）、中位数（50%）、75\%分位数（75%）、最大值（max），将统计好的结果赋值给变量df_summary_2，（前文括号内为各列的列名）

In [92]:
df_summary_2=df1.describe().T
df_summary_2

count       mean        std        min        25%        50%  \
指标A      82.0  34.050327  26.628904 -19.659707  10.828481  36.026232   
指标B      83.0  24.805453  28.622891 -19.874914  -0.024177  28.520651   
指标D      83.0  33.119837  28.715083 -18.120622   9.920332  34.768454   
指标E_累计   82.0  28.887801  27.812259 -19.444297   8.709318  29.143887   
指标F      83.0  32.348163  28.746736 -19.729022   9.377606  32.029436   
指标G      84.0  25.371621  27.835972 -24.793919   3.760644  26.529131   

              75%        max  
指标A     51.720217  79.181506  
指标B     48.255679  78.967743  
指标D     58.234900  79.788979  
指标E_累计  50.965432  78.368479  
指标F     60.092774  76.314237  
指标G     52.306199  74.853950

<strong>3.  </strong>

In [93]:
d1=df_summary_1[:2]
d2=df_summary_1[3:7]
df_summary_1=pd.concat([d1,d2])


In [94]:
df_summary=pd.concat([df_summary_2,df_summary_1],axis=1)
df_summary

count       mean        std        min        25%        50%  \
指标A      82.0  34.050327  26.628904 -19.659707  10.828481  36.026232   
指标B      83.0  24.805453  28.622891 -19.874914  -0.024177  28.520651   
指标D      83.0  33.119837  28.715083 -18.120622   9.920332  34.768454   
指标E_累计   82.0  28.887801  27.812259 -19.444297   8.709318  29.143887   
指标F      83.0  32.348163  28.746736 -19.729022   9.377606  32.029436   
指标G      84.0  25.371621  27.835972 -24.793919   3.760644  26.529131   

              75%        max      BegDt      EndDt  Obs  NaNNum  NaNPct  \
指标A     51.720217  79.181506 2015-01-01 2022-09-01   93       2    2.15   
指标B     48.255679  78.967743 2013-11-01 2022-09-01  107       1    0.93   
指标D     58.234900  79.788979 2013-11-01 2022-09-01  107       1    0.93   
指标E_累计  50.965432  78.368479 2013-11-01 2022-09-01  107       2    1.87   
指标F     60.092774  76.314237 2013-11-01 2022-09-01  107       1    0.93   
指标G     52.306199  74.853950 2013-11-01 2022-09-01  107       0    0.00   

        NegNum  NegPct                                             NegLst  
指标A         13   13.98  [2016-08, 2016-09, 2017-02, 2017-03, 2018-08, ...  
指标B         28   26.17  [2013-11, 2013-12, 2014-04, 2014-06, 2014-08, ...  
指标D         18   16.82  [2014-01, 2014-02, 2014-07, 2014-09, 2014-12, ...  
指标E_累计      20   18.69  [2013-11, 2014-03, 2014-06, 2015-01, 2015-06, ...  
指标F         19   17.76  [2014-07, 2015-01, 2015-11, 2016-07, 2016-11, ...  
指标G         27   25.23  [2014-01, 2014-03, 2014-07, 2014-08, 2014-12, ...

- Question 4

<strong>1.  </strong>请使用你认为的最简洁的方式：对于数据框df1中的各指标，使用各指标上一期的值对当期为缺失的进行填充，并将填补后的数据框赋值给变量df2

In [95]:
df2=df1.fillna(df1.median())
df2


指标A        指标B        指标D     指标E_累计        指标F        指标G
YYYYMM                                                                      
2015-01-01  45.434917 -12.905679  24.022993  -6.834541  -5.296895  52.968511
2015-02-01  31.919156  60.259495  14.858841  40.628302  45.376438  37.912412
2015-03-01  78.077235 -15.167676   7.043763  39.775122  53.367137  52.085428
2015-04-01  49.384843  28.609080  34.357724   8.668052  50.794603  39.574971
2015-05-01   7.336568  56.382674  34.768454  34.298297  15.006340  26.645740
...               ...        ...        ...        ...        ...        ...
2021-08-01   0.147898  44.543486  10.557716  29.407784 -16.316181  38.094446
2021-09-01  43.046844  30.111458  49.393706  54.065227  68.291341  41.396381
2021-10-01 -19.659707  29.809855  79.009139  -6.919456   3.515089  54.807505
2021-11-01  46.220272  39.486479   1.283920  74.777369   1.204036  -0.061766
2021-12-01  33.491257 -11.082423  41.962053  33.863632  33.335368  31.786463

[84 rows x 6 columns]

<strong>2.  </strong>
请使用**你认为的最简洁**的方式：由于各列指标中包含一些负值，故需要对数据框`df2`中的各指标逐列进行正向化操作（**Min-Max normalization**），使用的公式为：

$$
x' = \dfrac{x - \min \{ x  \}}{\max \{x\} - \min \{ x \}} + 1
$$

    并将正向化后的数据框赋值给变量`df3`

In [96]:
df3=(df2-df2.min())/(df2.max()-df2.min())+1
df3


指标A       指标B       指标D    指标E_累计       指标F       指标G
YYYYMM                                                                
2015-01-01  1.658578  1.070508  1.430434  1.128917  1.150267  1.780372
2015-02-01  1.521836  1.810727  1.336836  1.614159  1.677876  1.629279
2015-03-01  1.988828  1.047624  1.257017  1.605436  1.761075  1.771510
2015-04-01  1.698540  1.490517  1.535988  1.287410  1.734290  1.645964
2015-05-01  1.273128  1.771505  1.540183  1.549444  1.361664  1.516214
...              ...       ...       ...       ...       ...       ...
2021-08-01  1.200398  1.651727  1.292906  1.499445  1.035534  1.631106
2021-09-01  1.634417  1.505717  1.689558  1.751533  1.916466  1.664242
2021-10-01  1.000000  1.502665  1.992035  1.128049  1.242017  1.798827
2021-11-01  1.666523  1.600565  1.198188  1.963286  1.217954  1.248195
2021-12-01  1.537741  1.088954  1.613655  1.545000  1.552505  1.567803

[84 rows x 6 columns]

<strong>3.  </strong>请使用**你认为的最简洁**的方式：使用数据框`df3`，先计算得到各指标每月的**同比变动**，然后使用`pandas`**处理缺失值的相关函数**剔除包含`NaN`的月份，并将处理好的数据框赋值给变量`df_pct`
    - Note
        - 指标同比变动的计算公式为：$\dfrac{某指标今年的当期值}{该指标上一年度的当期值} - 1$

In [97]:
df_pct=(df3/df3.shift(12)-1).dropna(axis=0)
df_pct.index=pd.Series(df_pct.index.values).dt.to_period('M')

df_pct


指标A       指标B       指标D    指标E_累计       指标F       指标G
2016-01 -0.032286  0.565458  0.278633  0.154220  0.250641 -0.253840
2016-02  0.296626 -0.392276  0.494611 -0.346074 -0.179045 -0.354235
2016-03 -0.327860  0.487129 -0.052345  0.017573  0.054775  0.027381
2016-04  0.136067 -0.222924  0.191439  0.337606 -0.139896  0.110220
2016-05  0.493070  0.128984  0.228777 -0.032351  0.365893  0.194376
...           ...       ...       ...       ...       ...       ...
2021-08 -0.228822 -0.155668  0.010173 -0.168681 -0.482233 -0.068653
2021-09 -0.180598 -0.072508 -0.146516  0.172240  0.207649 -0.116894
2021-10 -0.293049  0.444849  0.811337 -0.024133 -0.365235  0.663010
2021-11 -0.032874  0.431046 -0.279621  0.364505  0.090800 -0.162298
2021-12  0.490234  0.088954  0.014964  0.262768 -0.069609  0.026491

[72 rows x 6 columns]

- Question 5

请使用**你认为的最简洁**的方式：依次保存以上作答过程中的数据框变量`[df_pct, df3, df2, df1, df_summary, df_temp, df]`至`Output`文件夹中的`Result.xlsx`的不同张工作簿
    - 保存时，Excel文件中的各张工作薄的名称依次对应为
        - `df_pct`：同比变动
        - `df3`：正向化后数据
        - `df2`：原始筛选数据(填补缺失后)
        - `df1`：原始筛选数据(填补缺失前)
        - `df_summary`：指标统计性描述
        - `temp_1`：观测月
        - `df`：原始数据

  

In [98]:
df_pct.index=df_pct.index.astype(str).str.replace("-","/")
df3.index=pd.Series(df3.index.values).dt.to_period('M').astype(str).str.replace("-","/")
df2.index=pd.Series(df2.index.values).dt.to_period('M').astype(str).str.replace("-","/")
df1.index=pd.Series(df1.index.values).dt.to_period('M').astype(str).str.replace("-","/")
df.index=pd.Series(df.index.values).dt.to_period('M').astype(str).str.replace("-","/")
writer = pd.ExcelWriter('../Output/Result.xlsx')
df_pct.to_excel(writer,sheet_name="同比变动")
df3.to_excel(writer,sheet_name="正向化后数据")
df2.to_excel(writer,sheet_name="原始筛选数据(填补缺失后)")
df1.to_excel(writer,sheet_name="原始筛选数据(填补缺失前)")
df_summary.to_excel(writer,sheet_name="指标统计性描述")
temp_1.to_excel(writer,sheet_name="观测月")
df.to_excel(writer,sheet_name="原始数据")
writer.save()
